src/indexer/merger.rs

```diff
@@ -520,6 +525,71 @@ impl IndexMerger {
+        serializer: &mut SegmentSerializer,
+        doc_id_mapping: &SegmentDocIdMapping,
+    ) -> crate::Result<()> {
+        use crate::spatial::bkd::Segment;
+        let mut segment_mappings: Vec<Vec<Option<DocId>>> = Vec::new();
+        for reader in &self.readers {
+            let max_doc = reader.max_doc();
+            segment_mappings.push(vec![None; max_doc as usize]);
+        }
+        for (new_doc_id, old_doc_addr) in doc_id_mapping.iter_old_doc_addrs().enumerate() {
+            segment_mappings[old_doc_addr.segment_ord as usize][old_doc_addr.doc_id as usize] =
+                Some(new_doc_id as DocId);
+        }
+        let mut temp_files: HashMap<Field, NamedTempFile> = HashMap::new();
+
+        for (field, field_entry) in self.schema.fields() {
+            if matches!(field_entry.field_type(), FieldType::Spatial(_)) {
+                temp_files.insert(field, NamedTempFile::new()?);
+            }
+        }
+        for (segment_ord, reader) in self.readers.iter().enumerate() {
+            for (field, temp_file) in &mut temp_files {
+                let spatial_readers = reader.spatial_fields();
+                let spatial_reader = match spatial_readers.get_field(*field)? {
+                    Some(reader) => reader,
+                    None => continue,
+                };
+                let segment = Segment::new(spatial_reader.get_bytes());
+                for triangle_result in LeafPageIterator::new(&segment) {
+                    let triangles = triangle_result?;
+                    for triangle in triangles {
+                        if let Some(new_doc_id) =
+                            segment_mappings[segment_ord][triangle.doc_id as usize]
+                        {
+                            for &word in &triangle.words {
+                                temp_file.write_all(&word.to_le_bytes())?;
+                            }
+                            temp_file.write_all(&new_doc_id.to_le_bytes())?;
+                        }
+                    }
+                }
+            }
+        }
+        if let Some(mut spatial_serializer) = serializer.extract_spatial_serializer() {
+            for (field, mut temp_file) in temp_files {
+                // Flush and sync triangles.
+                temp_file.flush()?;
+                temp_file.as_file_mut().sync_all()?;
+                // Memory map the triangle file.
+                use memmap2::MmapOptions;
+                let mmap = unsafe { MmapOptions::new().map_mut(temp_file.as_file())? };
+                // Cast to &[Triangle] slice
+                let triangle_count = mmap.len() / std::mem::size_of::<Triangle>();
+                let triangles = unsafe {
+                    std::slice::from_raw_parts_mut(mmap.as_ptr() as *mut Triangle, triangle_count)
+                };
+                // Get spatial writer and rebuild block kd-tree.
+                spatial_serializer.serialize_field(field, triangles)?;
+            }
```

Comment:
FYI  I imported your PR in https://github.com/quickwit-oss/tantivy/pull/2755 and started reviewing.

The test do not pass in your PR mainly due to :


```suggestion
            }
            spatial_serializer.close()?;
```